### PR TITLE
add encrypt/decrypt hsm API for storage security

### PR DESCRIPTION
**Pull request by LucasLi1024**

### CODE REVIEW

File: bcos-crypto/bcos-crypto/signature/hsmSM2/HsmSM2Crypto.cpp [hunk 1]
```diff
@@ -166,7 +168,7 @@ std::pair<bool, bytes> HsmSM2Crypto::recoverAddress(Hash::Ptr _hashImpl, bytesCo
         h256 r;
         h256 s;
     } in;
-    memcpy(&in, _input.data(), std::min(_input.size(), sizeof(_input)));
+    memcpy(&in, _input.data(), _input.size());
```
Comment (bxq2011hust) [inline bcos-crypto/bcos-crypto/signature/hsmSM2/HsmSM2Crypto.cpp:171]:
```suggestion
    memcpy(&in, _input.data(), std::min(_input.size(), sizeof(in)));
```

Comment (LucasLi1024) [reply bcos-crypto/bcos-crypto/signature/hsmSM2/HsmSM2Crypto.cpp]: fixed

File: bcos-gateway/bcos-gateway/GatewayFactory.cpp [hunk 2]
```diff
@@ -392,7 +392,8 @@ std::shared_ptr<boost::asio::ssl::context> GatewayFactory::buildSSLContext(
         try
         {
             if (nullptr == m_dataEncrypt)  // storage_security.enable = false
-                enNodeKeyContent = readContents(boost::filesystem::path(_smCertConfig.enNodeKey));
+                enNodeKeyContent =
```
Comment (bxq2011hust) [inline bcos-gateway/bcos-gateway/GatewayFactory.cpp:395]:
```suggestion
                {
                    enNodeKeyContent = readContents(boost::filesystem::path(_smCertConfig.enNodeKey));
                }
```

File: bcos-rpc/bcos-rpc/RpcFactory.cpp [hunk 3]
```diff
@@ -269,7 +269,8 @@ std::shared_ptr<bcos::boostssl::ws::WsConfig> RpcFactory::initConfig(
             try
             {
                 if (nullptr == m_dataEncrypt)  // storage_security.enable = false
-                    keyContent = readContents(boost::filesystem::path(_nodeConfig->enSmNodeKey()));
+                    keyContent =
+                    readContents(boost::filesystem::path(_nodeConfig->enSmNodeKey()));
```
Comment (bxq2011hust) [inline bcos-rpc/bcos-rpc/RpcFactory.cpp:273]: same as before

File: bcos-gateway/bcos-gateway/GatewayFactory.cpp (hunk 2 above)
Comment (LucasLi1024) [reply bcos-gateway/bcos-gateway/GatewayFactory.cpp]: fixed

File: bcos-rpc/bcos-rpc/RpcFactory.cpp (hunk 3 above)
Comment (LucasLi1024) [reply bcos-rpc/bcos-rpc/RpcFactory.cpp]: fixed

File: bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp [hunk 4]
```diff
@@ -72,5 +72,78 @@ bcos::bytesPointer HsmSM4Crypto::HsmSM4Decrypt(const unsigned char* _cipherData,
...
+    return encryptedData;
+}
+
+bcos::bytesPointer HsmSM4Crypto::HsmSM4DecryptWithInternalKey(
+    const unsigned char* _cipherData, size_t _cipherDataSize, const unsigned int _keyIndex)
+{
+    // iv data
+    bytes ivData(SM4_IV_DATA_SIZE);
+    memcpy(ivData.data(), _cipherData, SM4_IV_DATA_SIZE);
+    std::string ivStr(ivData.begin(), ivData.end());
```
Comment (bxq2011hust) [inline bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp:126]:
```suggestion

```

File: bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp [hunk 5]
```diff
@@ -72,5 +72,78 @@ bcos::bytesPointer HsmSM4Crypto::HsmSM4Decrypt(const unsigned char* _cipherData,
...
+}
+
+bcos::bytesPointer HsmSM4Crypto::HsmSM4EncryptWithInternalKey(const unsigned char* _plainData,
+    size_t _plainDataSize, const unsigned int _keyIndex, const unsigned char* _ivData,
+    size_t _ivDataSize)
+{
+    // iv data
+    bytes inputIvData(SM4_IV_DATA_SIZE);
+    memcpy(inputIvData.data(), _ivData, SM4_IV_DATA_SIZE);
+    std::string originIvStr(inputIvData.begin(), inputIvData.end());
```
Comment (bxq2011hust) [inline bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp:85]:
```suggestion
 
```

File: bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp [hunk 6]
```diff
@@ -72,5 +72,78 @@ bcos::bytesPointer HsmSM4Crypto::HsmSM4Decrypt(const unsigned char* _cipherData,
...
+    auto encryptCode = provider.EncryptWithInternalKey((unsigned int)_keyIndex, SM4_CBC,
+        (unsigned char*)_ivData, (unsigned char*)inDataV.data(), inDataVLen,
+        (unsigned char*)encryptedData->data(), &size);
+    if (encryptCode != SDR_OK)
+    {
+        CRYPTO_LOG(ERROR) << "[HsmSM4Crypto::HsmSM4EncryptWithInternalKey] encrypt ERROR "
+                          << LOG_KV("error", provider.GetErrorMessage(encryptCode));
+        BOOST_THROW_EXCEPTION(std::runtime_error("Hsm SM4 EncryptWithInternalKey error"));
+    }
+    CRYPTO_LOG(DEBUG) << "[HsmSM4Crypto::EncryptWithInternalKey] Encrypt Success";
```
Comment (bxq2011hust) [inline bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp:109]:
```suggestion

```

File: bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp [hunk 7]
```diff
@@ -72,5 +72,78 @@ bcos::bytesPointer HsmSM4Crypto::HsmSM4Decrypt(const unsigned char* _cipherData,
...
+    auto decryptCode = provider.DecryptWithInternalKey((unsigned int)_keyIndex, SM4_CBC,
+        ivData.data(), (unsigned char*)pureCipherData.data(), pureCipherDataSize,
+        (unsigned char*)decryptedData->data(), &size);
+    if (decryptCode != SDR_OK)
+    {
+        CRYPTO_LOG(ERROR) << "[HsmSM4Crypto::HsmSM4DecryptWithInternalKey] decrypt ERROR "
+                          << LOG_KV("error", provider.GetErrorMessage(decryptCode));
+        BOOST_THROW_EXCEPTION(std::runtime_error("Hsm SM4 DecryptWithInternalKey error"));
+    }
+    CRYPTO_LOG(DEBUG) << "[HsmSM4Crypto::DecryptWithInternalKey] Decrypt Success";
```
Comment (bxq2011hust) [inline bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp:146]:
```suggestion
 
```

File: bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.h [hunk 8]
```diff
@@ -19,19 +19,25 @@
  * @author lucasli
  */
 #pragma once
+
 #include <bcos-crypto/interfaces/crypto/SymmetricEncryption.h>
+#include <mutex>
```
Comment (bxq2011hust) [inline bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.h:24]:
```suggestion

```

File: bcos-tool/bcos-tool/NodeConfig.cpp [hunk 9]
```diff
@@ -580,6 +582,16 @@ void NodeConfig::loadStorageSecurityConfig(boost::property_tree::ptree const& _p
     {
         return;
     }
+
+    if (m_hsmEnable)
```
Comment (bxq2011hust) [inline bcos-tool/bcos-tool/NodeConfig.cpp:586]:
```suggestion
    if (m_enableHsm)
```

File: bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp (hunk 4 above)
Comment (LucasLi1024) [reply bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp:126]: has deleted

File: bcos-tool/bcos-tool/NodeConfig.cpp (hunk 9 above)
Comment (LucasLi1024) [reply bcos-tool/bcos-tool/NodeConfig.cpp]: fixed

File: bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.h (hunk 8 above)
Comment (LucasLi1024) [reply bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.h]: fixed

File: bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp (hunk 7 above)
Comment (LucasLi1024) [reply bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp]: fixed

File: bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp (hunk 6 above)
Comment (LucasLi1024) [reply bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp]: fixed

File: bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp (hunk 5 above)
Comment (LucasLi1024) [reply bcos-crypto/bcos-crypto/encrypt/HsmSM4Crypto.cpp]: fixed

File: bcos-crypto/bcos-crypto/encrypt/AESCrypto.cpp [hunk 10]
```diff
@@ -38,31 +38,44 @@ bytesPointer bcos::crypto::AESEncrypt(const unsigned char* _plainData, size_t _p
...
+    // get ivData from _cipherData
+    bytes ivData(AES_IV_DATA_SIZE);
+    memcpy(ivData.data(), _cipherData, AES_IV_DATA_SIZE);
+
+    // pure cipher data(exclude iv data)
+    size_t pureCipherDataSize = _cipherDataSize - (size_t)AES_IV_DATA_SIZE;
+    bytes pureCipherData(pureCipherDataSize);
+    memcpy(pureCipherData.data(), _cipherData + AES_IV_DATA_SIZE, pureCipherDataSize);
+
+    CInputBuffer ciper{(const char*)pureCipherData.data(), pureCipherDataSize};
```
Comment (bxq2011hust) [inline bcos-crypto/bcos-crypto/encrypt/AESCrypto.cpp:68]:
```suggestion
    CInputBuffer ciper{(const char*)(pureCipherData.data()+AES_IV_DATA_SIZE), pureCipherDataSize};
```

File: bcos-crypto/bcos-crypto/encrypt/AESCrypto.cpp [hunk 11]
```diff
@@ -38,31 +38,44 @@ bytesPointer bcos::crypto::AESEncrypt(const unsigned char* _plainData, size_t _p
...
 
     FixedBytes<AES_KEY_SIZE> fixedKeyData(_key, _keySize);
     CInputBuffer key{(const char*)fixedKeyData.data(), AES_KEY_SIZE};
 
-    FixedBytes<AES_IV_DATA_SIZE> fixedIVData(_ivData, _ivDataSize);
+    FixedBytes<AES_IV_DATA_SIZE> fixedIVData(
+        (unsigned char*)ivData.data(), (size_t)AES_IV_DATA_SIZE);
```
Comment (bxq2011hust) [inline bcos-crypto/bcos-crypto/encrypt/AESCrypto.cpp:74]:
```suggestion
        (unsigned char*)_cipherData, (size_t)AES_IV_DATA_SIZE);
```

File: bcos-security/bcos-security/DataEncryption.cpp [hunk 12]
```diff
@@ -142,9 +144,13 @@ std::shared_ptr<bytes> DataEncryption::decryptFile(const std::string& filename)
 
 std::string DataEncryption::encrypt(const std::string& data)
 {
+    random_bytes_engine rbe;
```
Comment (bxq2011hust) [inline bcos-security/bcos-security/DataEncryption.cpp:147]: 1. 兼容3.3以前的版本
2. iv要在这里存下来

File: bcos-security/bcos-security/HsmDataEncryption.cpp [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    {
+        BCOS_LOG(ERROR) << LOG_DESC("[HsmDataEncryption][ENCFILE] EncryptedFile error")
+                        << LOG_KV("what", boost::diagnostic_information(e));
+        BOOST_THROW_EXCEPTION(EncryptedFileError());
+    }
+
+    return decFileBytes;
+}
+
+std::string HsmDataEncryption::encrypt(const std::string& data)
```
Comment (bxq2011hust) [inline bcos-security/bcos-security/HsmDataEncryption.cpp:107]:
```suggestion
std::string HsmDataEncryption::encrypt(const std::string& data)
{
return encrypt(data.data(), data.size());
}
std::string HsmDataEncryption::encrypt(void* data, size_t size);
```

File: bcos-crypto/bcos-crypto/encrypt/AESCrypto.cpp (hunk 10 above)
Comment (LucasLi1024) [reply bcos-crypto/bcos-crypto/encrypt/AESCrypto.cpp]: fixed

File: bcos-crypto/bcos-crypto/encrypt/AESCrypto.cpp (hunk 11 above)
Comment (LucasLi1024) [reply bcos-crypto/bcos-crypto/encrypt/AESCrypto.cpp]: fixed

File: bcos-security/bcos-security/DataEncryption.cpp (hunk 12 above)
Comment (LucasLi1024) [reply bcos-security/bcos-security/DataEncryption.cpp]: fixed

File: bcos-security/bcos-security/HsmDataEncryption.cpp (hunk 13 above)
Comment (LucasLi1024) [reply bcos-security/bcos-security/HsmDataEncryption.cpp]: fixed

File: bcos-security/bcos-security/DataEncryption.cpp [hunk 14]
```diff
@@ -38,39 +39,42 @@ namespace bcos
...
         KeyCenter keyClient;
         keyClient.setIpPort(keyCenterIp, keyCenterPort);
-        m_dataKey = asString(keyClient.getDataKey(cipherDataKey, smCryptoType));
+        m_dataKey = asString(keyClient.getDataKey(cipherDataKey, m_nodeConfig->smCryptoType()));
 
         BCOS_LOG(INFO) << LOG_BADGE("DataEncryption::init") << LOG_KV("key_center_ip:", keyCenterIp)
                        << LOG_KV("key_center_port:", keyCenterPort);
     }
-
-    if (false == smCryptoType)
```
Comment (bxq2011hust) [inline bcos-security/bcos-security/DataEncryption.cpp:59]: check this

File: bcos-security/bcos-security/DataEncryption.cpp [hunk 15]
```diff
@@ -91,9 +96,24 @@ std::shared_ptr<bytes> DataEncryption::decryptContents(const std::shared_ptr<byt
...
+            static_cast<uint32_t>(bcos::protocol::BlockVersion::V3_3_VERSION))
+        {
+            random_bytes_engine rbe;
+            std::vector<unsigned char> ivData(16);
+            std::generate(std::begin(ivData), std::end(ivData), std::ref(rbe));
+
+            decFileBytesBase64Ptr = m_symmetricEncrypt->symmetricDecrypt(
+                (const unsigned char*)encFileBytes.data(), encFileBytes.size(),
+                (const unsigned char*)m_dataKey.data(), m_dataKey.size(), ivData.data(), 16);
+            decFileBytesBase64Ptr->push_back(*(ivData.data()));
```
Comment (bxq2011hust) [inline bcos-security/bcos-security/DataEncryption.cpp:109]:
```suggestion
            decFileBytesBase64Ptr->insert();
```

File: bcos-security/bcos-security/DataEncryption.cpp [hunk 16]
```diff
@@ -115,17 +135,30 @@ std::shared_ptr<bytes> DataEncryption::decryptFile(const std::string& filename)
...
```
Comment (bxq2011hust) [inline bcos-security/bcos-security/DataEncryption.cpp:154]: fix

File: bcos-security/bcos-security/HsmDataEncryption.cpp [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+    std::generate(std::begin(ivData), std::end(ivData), std::ref(rbe));
+    // iv data would be changed after hsm encrypt, so keep it
+    auto originIvData = ivData;
+
+    bytesPointer encData = m_symmetricEncrypt->symmetricEncryptWithInternalKey(
+        reinterpret_cast<const unsigned char*>(data), size, m_encKeyIndex, ivData.data(),
+        SM4_IV_DATA_SIZE);
+    // append iv data to end of encData
+    encData->insert(encData->end(), originIvData.begin(), originIvData.end());
+    std::string value(encData->size(), 0);
```
Comment (bxq2011hust) [inline bcos-security/bcos-security/HsmDataEncryption.cpp:74]:
```suggestion
    std::string value((char*)encData->data(), encData->size());
    value.insert(encData->end(), originIvData.begin(), originIvData.end());
```

File: bcos-security/bcos-security/HsmDataEncryption.cpp [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+    return value;
+}
+
+std::string HsmDataEncryption::decrypt(uint8_t* data, size_t size)
+{
+    size_t offsetIv = size - SM4_IV_DATA_SIZE;
+    size_t cipherDataSize = size - SM4_IV_DATA_SIZE;
+    bytesPointer decData = m_symmetricEncrypt->symmetricDecryptWithInternalKey(
+        data, cipherDataSize, m_encKeyIndex, data + offsetIv, SM4_IV_DATA_SIZE);
+    std::string value(decData->size(), 0);
```
Comment (bxq2011hust) [inline bcos-security/bcos-security/HsmDataEncryption.cpp:86]:
```suggestion
    std::string value((char*)decData->data(), decData->size());
```

File: bcos-security/bcos-security/DataEncryption.cpp (hunk 15 above)
Comment (LucasLi1024) [reply bcos-security/bcos-security/DataEncryption.cpp:109]: fixed

**Review by bxq2011hust — APPROVED** [review]

File: bcos-security/bcos-security/HsmDataEncryption.cpp (hunk 18 above)
Comment (LucasLi1024) [reply bcos-security/bcos-security/HsmDataEncryption.cpp]: fixed

File: bcos-security/bcos-security/HsmDataEncryption.cpp (hunk 17 above)
Comment (LucasLi1024) [reply bcos-security/bcos-security/HsmDataEncryption.cpp]: fixed

File: bcos-security/bcos-security/DataEncryption.cpp (hunk 14 above)
Comment (LucasLi1024) [reply bcos-security/bcos-security/DataEncryption.cpp:59]: fixed

**Review by wenlinlee — APPROVED** [review]

**Review by JimmyShi22 — APPROVED** [review]

**Pull request merged**